gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -117,11 +130,10 @@ public void flush() throws InterruptedException {
 
   /** Sends accumulated elements asynchronously for batching. */
   private void sendBatch() {
-    if (currentOpenBatch == null) {
-      return;
-    }
     final Batch<ElementT, ElementResultT, RequestT, ResponseT> accumulatedBatch = currentOpenBatch;
```

Comment:
This has to be done under lock

Reply:
If I include this under lock then I would have to include callback inside the locks which we [tried to resolve](https://github.com/googleapis/gax-java/pull/744#discussion_r299078766) in earlier commits.

I think we have to keep callbacks inside the lock, Please let me know If I am missing something which can help me to resolve our problem.

Reply:
Not sure what you mean. To clarify what I said:

```java
final Batch<ElementT, ElementResultT, RequestT, ResponseT> batchToSend;

synchronized (elementLock) {
    if (currentOpenBatch == null) {	
      return;	
    }
    batchToSend = currentOpenBatch;
    currentOpenBatch = new Batch<>(prototype, batchingDescriptor, batchingSettings);
}
```

This is necessary because `sendBatch` is racing with itself, so the swap has to be atomic.

What were you thinking had to be included in the lock? What am I missing?

Reply:
Not sure what you mean. To clarify what I said:

```java
final Batch<ElementT, ElementResultT, RequestT, ResponseT> batchToSend;

synchronized (elementLock) {
    if (currentOpenBatch == null) {	
      return;	
    }
    batchToSend = currentOpenBatch;
    currentOpenBatch = new Batch<>(prototype, batchingDescriptor, batchingSettings);
}
```

This is necessary because `sendBatch` is racing with itself, so the swap has to be atomic.

What were you thinking had to be included in the lock? What am I missing?

Reply:
Also please add test to ensure this behavior:
 
- start 1 thread to continually add sequential elements to the batcher
- start 20 threads to call sendBatch a few thousand times
- use a fake unary callable to make sure that none of elements are repeated

This test should prove that the current code is incorrect

Reply:
Oops I accidentally deleted my other comment

To clarify my original comment about expanding the lock:

```java
final Batch<ElementT, ElementResultT, RequestT, ResponseT> batchToSend;

synchronized (elementLock) {
      if (currentOpenBatch == null) {	
      return;	
    }
  batchToSend = currentOpenBatch;
  currentOpenBatch = new Batch<>(prototype, batchingDescriptor, batchingSettings);
}

The swap has to happen atomically. Please clarify what you meant about needing to expand the lock to include callbacks in case I'm missing something
```

Reply:
Also, I just realized, that currentOpenBatch is never null and you need to check if its empty. Please add a test to make sure that empty batches are never sent

Reply:
Thanks a lot for the detailed instructions & test case scenario.
I have updated this PR with the above changes. Please have a look.